src/error_handling.c

```diff
@@ -14,313 +14,161 @@
-        "  '%s'\n"
-        "\n"
-        "with this new error message:\n"
-        "\n"
-        "  '%s, at %s:%zu'\n"
-        "\n"
-        "rcutils_reset_error() should be called after error handling to avoid this.\n"
-        "<<<\n",
-        old_error_string,
-        error_string,
-        file,
-        line_number);
-    }
-#endif
-    __rcutils_reset_error(&old_error_state);
-  }
-  __rcutils_reset_error_string(&__rcutils_error_string, __rcutils_error_state->allocator);
-}
+  // forcing the values back to their initial state should force the thread-local storage
+  // to initialize and do any required memory allocation
+  gtls_rcutils_thread_local_initialized = true;
+  rcutils_reset_error();
+  RCUTILS_SET_ERROR_MSG("no error - initializing thread-local storage");
+  rcutils_error_string_t throw_away = rcutils_get_error_string();
+  (void)throw_away;
+  rcutils_reset_error();
 
-const rcutils_error_state_t *
-rcutils_get_error_state(void)
-{
-#ifdef RCUTILS_THREAD_LOCAL_PTHREAD
-  return (rcutils_error_state_t *)pthread_getspecific(__rcutils_error_state_key);
-#else
-  return __rcutils_error_state;
-#endif
+  // at this point the thread-local allocator, error state, and error string are all initialized
+  return RCUTILS_RET_OK;
 }
 
-static void
-format_error_string(void)
+static
+bool
+__same_string(const char * str1, const char * str2, size_t count)
 {
-#ifdef RCUTILS_THREAD_LOCAL_PTHREAD
-  rcutils_error_state_t * __rcutils_error_state =
-    (rcutils_error_state_t *)pthread_getspecific(__rcutils_error_state_key);
-  char * __rcutils_error_string = (char *)pthread_getspecific(__rcutils_error_string_key);
-#endif
-  if (!__rcutils_error_is_set(__rcutils_error_state)) {
-    return;
-  }
-  size_t bytes_it_would_have_written = snprintf(
-    NULL, 0,
-    __error_format_string,
-    __rcutils_error_state->message,
-    __rcutils_error_state->file,
-    __rcutils_error_state->line_number);
-  rcutils_allocator_t allocator = __rcutils_error_state->allocator;
-  __rcutils_error_string =
-    (char *)allocator.allocate(bytes_it_would_have_written + 1, allocator.state);
-#ifdef RCUTILS_THREAD_LOCAL_PTHREAD
-  pthread_setspecific(__rcutils_error_string_key, __rcutils_error_string);
-#endif
-  if (NULL == __rcutils_error_string) {
-#if RCUTILS_REPORT_ERROR_HANDLING_ERRORS
-    // rcutils_allocate failed, but fwrite might work?
-    RCUTILS_SAFE_FWRITE_TO_STDERR(
-      "[rcutils|error_handling.c:" RCUTILS_STRINGIFY(__LINE__)
-      "] failed to allocate memory for the error string\n");
-#endif
-    return;
-  }
-  snprintf(
-    __rcutils_error_string, bytes_it_would_have_written + 1,
-    __error_format_string,
-    __rcutils_error_state->message,
-    __rcutils_error_state->file,
-    __rcutils_error_state->line_number);
-  // The Windows version of snprintf does not null terminate automatically in all cases.
-  __rcutils_error_string[bytes_it_would_have_written] = '\0';
+  return str1 == str2 || 0 == strncmp(str1, str2, count);
```

Comment:
Curious, the following cope compiles and outputs "true" for me with gcc and g++:

```c
#include <stdio.h>

int mycompare(const char * foo, const char * bar)
{
  return foo == bar;
}

int main(void)
{
  const char * foo = NULL;
  const char * bar = NULL;
  if (mycompare(foo, bar)) {
    printf("true\n");
  }
  else {
    printf("false\n");
  }
  return 0;
}
```

Reply:
But I see what you mean; if they are not initialized as `char *`.

Reply:
You'd have to do this to be equivalent to the code in the repository:

```c++
typedef struct my_string_t { char str[10]; } my_string_t;

int mycompare(const char * foo, const char * bar)
{
  return foo == bar;
}

int main(void)
{
  my_string_t foo = {.str = "test"};
  my_string_t bar = {.str = "test"};
  if (mycompare(foo.str, bar.str)) {
    printf("true\n");
  } else {
    printf("false\n");
  }
  return 0;
}
```

In which case the comparison doesn't produce a warning, but they can also never be null. I could change the `__same_str` function to take the `rcutils_error_string_t` instead.

Reply:
Sorry, I'm only half-right about this, the first argument is passed by the user, so it could be null.

I've found some things I'll refactor and test for here.

Reply:
> Sorry, I'm only half-right about this, the first argument is passed by the user, so it could be null.

As long as they're both not null.